src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -2156,36 +2229,93 @@ void InterpreterSelectQuery::executeTotalsAndHaving(
 void InterpreterSelectQuery::executeRollupOrCube(QueryPlan & query_plan, Modificator modificator)
 {
     const auto & header_before_transform = query_plan.getCurrentDataStream().header;
-
     ColumnNumbers keys;
-
-    for (const auto & key : query_analyzer->aggregationKeys())
-        keys.push_back(header_before_transform.getPositionByName(key.name));
+    ColumnNumbers all_keys;
+    ColumnNumbersTwoDimension keys_vector;
+    auto & query = getSelectQuery();
+    if (query.group_by_with_grouping_sets)
```

Comment:
These two fragment of code are not entirely the same since there are some parameter difference. I'm not sure if we need to move it to a separate function

Reply:
I do't see any problem to make these parameters function arguments.